general/hash.hpp

```diff
@@ -483,6 +495,36 @@ void HashTable<T>::DeleteAll()
    unused.DeleteAll();
 }
 
+template<typename T>
+void HashTable<T>::Alloc(int id, int p1, int p2)
+{
+   // enlarge the BlockArray to hold 'id'
+   while (id >= Base::Size())
+   {
+      Base::At(Base::Append()).next = -2; // append "unused" items
+   }
+
+   T& item = Base::At(id);
+   if (item.next == -2)
+   {
+      item.next = -1;
+      item.p1 = p1;
+      item.p2 = p2;
+
+      Insert(Hash(p1, p2), id, item);
+   }
+}
+
+template<typename T>
+void HashTable<T>::UpdateUnused()
+{
+   unused.DeleteAll();
+   for (int i = 0; i < Base::Size(); i++)
+   {
+      if (Base::At(i).next == -2) { unused.Append(i); }
+   }
+}
+
```

Comment:
Is item.next=-1 needed here?

Also, is UpdateUnused just "Reset" the HashTable

Reply:
No, this just updates the `unused` list. This is a special function that's only needed when repopulating the table e.g. when loading from a file.